### DISCUSSION TITLE
Brain bug or Berry Bug?

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#20136] [Q&A] [answered]
hi,
i have the following code

```
# Basisklasse Sensor
class Sensor
    var id, name, sensortype, room, floor 
    static all=[]

    
    def init(name, sensortype, room, floor )
        self.id = id_counter_sensor
        id_counter_sensor += 1
        self.name = name
        self.sensortype = sensortype
        self.room = room
        self.floor = floor 
        Sensor.all.push(self)        
    end
    
    def triggerNo(ip_number) # wird  von parse_trigger_msg() aufgerufen
       print("||>> sensor", ip_number, "hat den Wert:XXX")
    end
    
end

```

If i enter this in Berry console , i got this behavier:
![grafik](https://github.com/arendst/Tasmota/assets/16272493/a8088835-d770-43b8-b2d0-39f74df3a590)

Could somone explain this?

**Comment by s-hadinger** [comment]
I'm confused. You are mixing instance values and global values (id_counter_sensor) which is a bad practice. Then you are not creating a single instance with this class, which makes it useless.

Then you are calling a method through a class identifier and not an instance method. You can't do that.

Either declare the method as `static def triggerNo(ip_number`... or create an instance of the class and call the method.

Internally when you declare a method like `def triggerNo(ip_number)`, it's actually equivalent to `static def triggerNo(self, ip_number)`. `self` is an implicit first argument. So in your example, `ip_number` is the second argument hence `nil` in the first call.

**Comment by ChrisVonFrosta** [comment]
i understand your confusion. That was only a snippet. A have a class Sensor and more than one subclass (like Buttons or PIRs).
All subclasses have a methon called "trigger()".

That is a try do make a method in mainclass to iterate all or hole subclasses.
For example: to find the correct Subclass instance, corresponding to the Input that was switched.

I am relatively new on object-oriented programming and Berry. So i try to find options ;)

Your last Info give me the reason for that behavier.

But i do not no what  the declaration of `static def def triggerNo(self, ip_number)` means.
I have to check.

sure, i could workaround by giving an extra parameter in first. But is there a better way?

thx alot!

**Reply by s-hadinger** [reply]
Sorry it's a typo. It's `static def triggerNo(self, ip_number)`

**Reply by s-hadinger** [reply, marked as answer]
Here is the documentation for `static` methods:

https://github.com/berry-lang/berry/wiki/Chapter-6#class-methods-static

**Reply by ChrisVonFrosta** [reply]
ah, so easy. just adding a "static in front did the job.
Very Nice! Tank you!!

**Comment by ChrisVonFrosta** [comment]
it seems to be an other or additional problem.
Sometimes, upload is totaly broken.  Than no files are uploaded. Either with promt cmd `curl -F "file=@C:\Dev\Tasmota\path\filename.be" "http://192.168.xx.xx/ufsu" `,
nor with the webinterface.
rebooting do not fix this reliable. I just needed 3 reboots to be able to upload files again. What kind of information can i offer?

**Reply by s-hadinger** [reply]
I have seen this problem before, indeed. But it's unrelated to Berry. Please open a new discussion and provide any relevant logs
